File: 2nd year/MPP/project-mpp/csharp-project/persistence/dbRepository/DbAthleteRepository.cs
using System;
using System.Collections.Generic;
using log4net;
using Microsoft.Data.Sqlite;
using model;
using persistence.validators;

namespace persistence.dbRepository
{
    public class DbAthleteRepository : IAthleteRepository
    {
        private static readonly ILog Log = LogManager.GetLogger(typeof(DbAthleteRepository));
        private readonly SqliteConnection _connection = JdbcManager.GetConnection();
        private readonly IValidator<Athlete> _validator;

        public DbAthleteRepository(IValidator<Athlete> validator)
        {
            this._validator = validator;
        }

        public Athlete FindOne(long id)
        {
            Log.Info($"Entered with id = {id}");
            try
            {
                const string commandText = "select * from Athlete where id = @id";
                using var command = new SqliteCommand(commandText, _connection);
                Log.Debug(commandText);

                command.Parameters.AddWithValue("id", id);

                using var reader = command.ExecuteReader();
                if (reader.Read())
                {
                    return new Athlete
                    {
                        Id = reader.GetInt64(0),
                        Name = reader.GetString(1)
                    };
                }
            }
            catch (Exception ex)
            {
                Log.Error(ex.Message);
                Console.Write(ex.Message);
            }

            return null;
        }

        public IEnumerable<Athlete> FindAll()
        {
            Log.Info("");
            const string commandText = "select * from Athlete";
            IList<Athlete> athletes = new List<Athlete>();

            try
            {
                using var command = new SqliteCommand(commandText, _connection);
                Log.Debug(commandText);

                using var reader = command.ExecuteReader();
                while (reader.Read())
                {
                    athletes.Add(new Athlete
                    {
                        Id = reader.GetInt64(0),
                        Name = reader.GetString(1)
                    });
                }
            }
            catch (Exception ex)
            {
                Log.Error(ex.Message);
                Console.Write(ex.Message);
            }

            Log.Info(athletes);
            return athletes;
        }

        public void Save(Athlete athlete)
        {
            Log.Info($"Entered with entity = {athlete}");
            const string commandText =
                "insert into Athlete (id, name) values (@id, @name)";
            _validator.Validate(athlete);
            try
            {
                using var command = new SqliteCommand(commandText, _connection);
                Log.Debug(commandText);

                command.Parameters.AddWithValue("id", athlete.Id);
                command.Parameters.AddWithValue("name", athlete.Name);

                command.ExecuteNonQuery();
            }
            catch (Exception ex)
            {
                Log.Error(ex.Message);
                Console.Write(ex.Message);
            }
        }

        public void Update(Athlete athlete)
        {
            Log.Info($"Entered with entity = {athlete}");
            const string commandText =
                "update Athlete set name = @name where id = @id";
            _validator.Validate(athlete);
            try
            {
                using var command = new SqliteCommand(commandText, _connection);
                Log.Debug(commandText);

                command.Parameters.AddWithValue("id", athlete.Id);
                command.Parameters.AddWithValue("name", athlete.Name);

                command.ExecuteNonQuery();
            }
            catch (Exception ex)
            {
                Log.Error(ex.Message);
                Console.Write(ex.Message);
            }
        }

        public void Delete(long id)
        {
            Log.Info($"Entered with id = {id}");
            const string commandText = "delete from Athlete where id = @id";
            try
            {
                using var command = new SqliteCommand(commandText, _connection);
                Log.Debug(commandText);

                command.Parameters.AddWithValue("id", id);

                command.ExecuteNonQuery();
            }
            catch (Exception ex)
            {
                Log.Error(ex.Message);
                Console.Write(ex.Message);
            }
        }
    }
}

File: 2nd year/MPP/project-mpp/java-project/TriathlonPersistence/src/main/java/triathlon/persistence/dbRepository/DBAthleteRepository.java
package triathlon.persistence.dbRepository;

import triathlon.model.Athlete;
import triathlon.persistence.validators.ValidationException;
import triathlon.persistence.RepositoryException;
import triathlon.persistence.AthleteRepository;
import triathlon.persistence.validators.Validator;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class DBAthleteRepository implements AthleteRepository {
    private final Validator<Athlete> validator;
    private final Connection connection;

    public DBAthleteRepository(Properties props, Validator<Athlete> validator) {
        this.connection = new JDBCManager(props).getConnection();
        this.validator = validator;
    }

    @Override
    public Athlete findOne(Long id) {
        final String sql_query = "select * from Athlete where id = ?";

        try (PreparedStatement statement = connection.prepareStatement(sql_query)) {
            statement.setLong(1, id);
            ResultSet resultSet = statement.executeQuery();
            if (resultSet.next()) {
                return new Athlete(
                        resultSet.getLong("id"),
                        resultSet.getString("name")
                );
            }
        } catch (SQLException e) {
            System.err.println("Error DB: " + e);
        }
        return null;
    }

    @Override
    public Iterable<Athlete> findAll() {
        final String sql_query = "select * from Athlete";

        try (PreparedStatement statement = connection.prepareStatement(sql_query)) {
            List<Athlete> athletes = new ArrayList<>();
            ResultSet resultSet = statement.executeQuery();
            while (resultSet.next()) {
                athletes.add(new Athlete(
                        resultSet.getLong("id"),
                        resultSet.getString("name")
                ));
            }
            return athletes;
        } catch (SQLException e) {
            System.err.println("Error DB: " + e);
        }

        return null;
    }

    @Override
    public void save(Athlete athlete) {
        final String sql_query = "insert into Athlete (name) values (?)";
        try {
            validator.validate(athlete);
        } catch (ValidationException ex) {
            throw new RepositoryException(ex.getMessage());
        }

        try (PreparedStatement statement = connection.prepareStatement(sql_query)) {
            statement.setString(1, athlete.getName());
            int updates = statement.executeUpdate();
        } catch (SQLException e) {
            System.err.println("Error DB: " + e);
        }
    }

    @Override
    public void update(Athlete athlete) {
        final String sql_query = "update Athlete set name = ? where id = ?";
        try {
            validator.validate(athlete);
        } catch (ValidationException ex) {
            throw new RepositoryException(ex.getMessage());
        }

        try (PreparedStatement statement = connection.prepareStatement(sql_query)) {
            statement.setString(1, athlete.getName());
            statement.setLong(2, athlete.getId());
            int updates = statement.executeUpdate();
        } catch (SQLException e) {
            System.err.println("Error DB: " + e);
        }
    }

    @Override
    public void delete(Long id) {
        final String sql_query = "delete from Athlete where id = ?";

        try (PreparedStatement statement = connection.prepareStatement(sql_query)) {
            statement.setLong(1, id);

            int updates = statement.executeUpdate();
        } catch (SQLException e) {
            System.err.println("Error DB: " + e);
        }
    }
}